src/Functions/FunctionsComparison.h

```diff
@@ -1216,7 +1216,10 @@ class FunctionComparison : public IFunction
         {
             return res;
         }
-        else if (isColumnedAsDecimal(left_type) || isColumnedAsDecimal(right_type))
+        else if ((isColumnedAsDecimal(left_type) || isColumnedAsDecimal(right_type))
+                 // Comparing Date and DateTime64 requires implicit conversion,
+                 // otherwise Date is treated as number.
+                 && !(date_and_datetime && (isDate(left_type) || isDate(right_type))))
```

Comment:
Ok but I would rather rename `date_and_datetime`, cause it may have 2 different datetimes